chore(RHTAPWATCH-552): sync staging member cluster's for segment-bridge

Omeramsc opened the pull request:
The segment-bridge component had sync issues on some of its object. The cause seem to be the `toolchain-host-operator` setting, which moved into the segment-bridge directory and so got applied to all of its cluster's settings.

Applying the settings to the RH clusters in which
the `toolchain-host-operator` component does not exist, seem to be the cause for the unsync issue.

This change separates the `toolchain-host-operator` settings between the different clusters for the segment-bridge component. currently applied only for the staging environment, to verify it works before moving to production.

hugares commented:
I maybe missing something here but how does the Application from the  ApplicationSet picks up the right folder of the component, e.g. components/segment-bridge/staging/stone-stg-m01?

```
kustomize build argo-cd-apps/overlays/staging | grep segment-bridge-member$ -A 29
  name: segment-bridge-member
  namespace: argocd-staging
spec:
  generators:
  - merge:
      generators:
      - clusters:
          selector:
            matchLabels:
              appstudio.redhat.com/member-cluster: "true"
          values:
            clusterDir: ""
            environment: staging
            sourceRoot: components/segment-bridge
      - list:
          elements: []
      mergeKeys:
      - nameNormalized
  template:
    metadata:
      name: segment-bridge-member-{{nameNormalized}}
    spec:
      destination:
        namespace: segment-bridge
        server: '{{server}}'
      project: default
      source:
        path: '{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'
        repoURL: https://github.com/redhat-appstudio/infra-deployments.git
        targetRevision: main
```
The path point to `'{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'` but `clusterDir` is not set.

ifireball commented on components/segment-bridge/base/kustomization.yaml:
```diff
@@ -1,5 +1,4 @@
 resources:
-- ./rbac
-- serviceaccount.yaml
+  - serviceaccount.yaml
```
the file name here make it really hard to understand that it sets up the service account for reading namespaces on member clusters, and not only the account but also its cluster role and role binding as well as a role and role binding for reading it by particular members of the O11y team.

We should probably re-factor this to multiple files with clearer names at some point.

Omeramsc replied:
out of scope, will be fixed in a separate change.

ifireball commented on components/segment-bridge/staging/stone-stg-host/kustomization.yaml:
```diff
@@ -1,5 +1,4 @@
 resources:
-- ../base
-
+  - ../../rbac
```
Similarly to the other comment "rbac" isn't really telling me its setting up a service account and associated roles for reading `Usersignup` resources from the `toolchain-host-operator` namespace on the host cluster.

Omeramsc replied:
The files are being moved around, not modified, to keep this change as small and atomic as possible so to fix the bug. I agree namings can be better, a separate ticket should be opened for it.

ifireball commented:
> I maybe missing something but how does the Application from the ApplicationSet picks up the right folder of the component, e.g. components/segment-bridge/staging/stone-stg-m01?
> 
> ```
> kustomize build argo-cd-apps/overlays/staging | grep segment-bridge-member$ -A 29
>   name: segment-bridge-member
>   namespace: argocd-staging
> spec:
>   generators:
>   - merge:
>       generators:
>       - clusters:
>           selector:
>             matchLabels:
>               appstudio.redhat.com/member-cluster: "true"
>           values:
>             clusterDir: ""
>             environment: staging
>             sourceRoot: components/segment-bridge
>       - list:
>           elements: []
>       mergeKeys:
>       - nameNormalized
>   template:
>     metadata:
>       name: segment-bridge-member-{{nameNormalized}}
>     spec:
>       destination:
>         namespace: segment-bridge
>         server: '{{server}}'
>       project: default
>       source:
>         path: '{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'
>         repoURL: https://github.com/redhat-appstudio/infra-deployments.git
>         targetRevision: main
> ```
> 
> The path point to `'{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'` but `clusterDir` is not set.

@Omeramsc, @hugares is right, you've reated directories with the cluster names but that does not provide a mechanism that would actually limit a resource to a particular kind of cluster. We have the `deploy-to-host-cluster` and `deploy-to-member-cluster` [k-components](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/1802-kustomize-components/README.md) to help us achieve that. You can find numerous examples of them being used in this repo.

Omeramsc commented:
> > I maybe missing something but how does the Application from the ApplicationSet picks up the right folder of the component, e.g. components/segment-bridge/staging/stone-stg-m01?
> > ```
> > kustomize build argo-cd-apps/overlays/staging | grep segment-bridge-member$ -A 29
> >   name: segment-bridge-member
> >   namespace: argocd-staging
> > spec:
> >   generators:
> >   - merge:
> >       generators:
> >       - clusters:
> >           selector:
> >             matchLabels:
> >               appstudio.redhat.com/member-cluster: "true"
> >           values:
> >             clusterDir: ""
> >             environment: staging
> >             sourceRoot: components/segment-bridge
> >       - list:
> >           elements: []
> >       mergeKeys:
> >       - nameNormalized
> >   template:
> >     metadata:
> >       name: segment-bridge-member-{{nameNormalized}}
> >     spec:
> >       destination:
> >         namespace: segment-bridge
> >         server: '{{server}}'
> >       project: default
> >       source:
> >         path: '{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'
> >         repoURL: https://github.com/redhat-appstudio/infra-deployments.git
> >         targetRevision: main
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The path point to `'{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'` but `clusterDir` is not set.
> 
> @Omeramsc, @hugares is right, you've reated directories with the cluster names but that does not provide a mechanism that would actually limit a resource to a particular kind of cluster. We have the `deploy-to-host-cluster` and `deploy-to-member-cluster` [k-components](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/1802-kustomize-components/README.md) to help us achieve that. You can find numerous examples of them being used in this repo.

I fixed the files and moved some things around, following other similar components; member clusters should not use the manifests for the host cluster's configurations.
as for the k-components,  `deploy-to-member-cluster-merge-generator` is already applied to segment-bridge's member cluster configurations; I did not applied the same for the host cluster configs as I couldn't find definitions for those components (or the difference between the "merge-generator" and non "merge-generator" components).

ifireball commented on argo-cd-apps/base/member/infra-deployments/pipeline-service/pipeline-service.yaml:
```diff
@@ -29,18 +29,18 @@ spec:
       project: default
       source:
-        path: '{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}'
+        path: "{{values.sourceRoot}}/{{values.environment}}/{{values.clusterDir}}"
```
why are we touching this file here?

Omeramsc replied:
Seem like auto-formatting of my IDE, will revert

hugares commented on argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge-member.yaml:
```diff
@@ -12,9 +12,13 @@ spec:
+                - nameNormalized: stone-stg-m01
+                  values.clusterDir: stone-stg-m01
+                - nameNormalized: stone-stg-rh01
+                  values.clusterDir: stone-stg-rh01
```
same comment as in `argo-cd-apps/base/host/segment-bridge/segment-bridge-host.yaml`

Omeramsc replied:
responded above

Omeramsc commented on argo-cd-apps/base/host/segment-bridge/segment-bridge-host.yaml:
```diff
@@ -12,9 +12,11 @@ spec:
-              elements: []
+              elements:
+                - nameNormalized: stone-stg-host
+                  values.clusterDir: stone-stg-host
```
Intresting, but won't it use all the directories under `staging`? meaning that the settings for the host cluster will include both`stone-stg-host` and `stone-stg-rh01` and `stone-sth-m01`, instead of the current setting where I can have it use `stone-stg-host` only for the host configurations, and the others for the member configurations?

hugares replied:
no it will not. This generators spec determine how to create application. The `clusters` generator will create one application per cluster which matches the criteria. You do not see the criteria in this file because it is added by another kustomization file. If you do the kustomize build, you will see the criteria:
```
hares@hares-thinkpadp1gen3 infra-deployments]$ kustomize build argo-cd-apps/overlays/staging/ | grep segment-bridge-host$ -A29
  name: segment-bridge-host
  namespace: argocd-staging
spec:
  generators:
  - merge:
      generators:
      - clusters:
          selector:
            matchLabels:
              appstudio.redhat.com/host-cluster: "true"
          values:
            clusterDir: base
            environment: staging
            sourceRoot: components/segment-bridge
      - list:
          elements:
          - nameNormalized: stone-stg-host
            values.clusterDir: stone-stg-host
      mergeKeys:
      - nameNormalized
```
The `clusters` will only create one application per cluster having `appstudio.redhat.com/host-cluster: "true"` label.

That said, maybe we should proceed with the change the way it is now as it is consistent with other components and refactoring could be done separately, for all components at once.

Omeramsc replied:
I see, OK. 
It's a very easy change, but since you already mentioned it - it's probably wise to wait for a repo-wide refactor for consistency.

hugares commented on argo-cd-apps/base/host/segment-bridge/kustomization.yaml:
```diff
@@ -4,4 +4,4 @@ resources:
 components:
   - ../../../k-components/inject-infra-deployments-repo-details
-
+  
```
nit: you added trailing space

Omeramsc replied:
fixed

hugares commented on argo-cd-apps/base/host/segment-bridge/segment-bridge-host.yaml:
```diff
@@ -39,4 +41,4 @@ spec:
             factor: 2
             maxDuration: 3m
-
+            
```
nit: you added trailing space

Omeramsc replied:
fixed

hugares commented on components/segment-bridge/base/kustomization.yaml:
```diff
@@ -1,5 +1,5 @@
-- ./rbac
-- serviceaccount.yaml
+  - serviceaccount.yaml
+  
```
nit: you added trailing space

Omeramsc replied:
fixed

hugares commented on components/segment-bridge/staging/base/kustomization.yaml:
```diff
@@ -1,5 +1,5 @@
-  - segment-bridge.yaml
\ No newline at end of file
+  - ../../base
+  
```
nit:  trailing spaces, same comment for other files below

Omeramsc replied:
fixed

hugares commented:
/lgtm

Omeramsc commented:
/retest
